### PR TITLE
Add UseVolatileReadWriteAnalyzer

### PR DESCRIPTION
**Pull request by CollinAlpert**
This PR adds an analyzer which flags invocations to `Thread.VolatileRead` and `Thread.VolatileWrite` and a fixer which rewrites the calls to `Volatile.Read` and `Volatile.Write` respectively.

Fixes dotnet/runtime#27997

### CODE REVIEW

**Comment by stephentoub** [comment]
> This PR adds an analyzer

Thanks, though I thought the plan was to mark them as obsolete and then have a fixer associated with that diagnostic id? Or did that plan change?

**Comment by CollinAlpert** [comment]
That was my understanding as well, however I couldn't figure out how to add tests without an analyzer emitting the diagnostic ID and I also thought it would benefit developers immediately instead of having to wait for the obsoletion of the APIs.

But I am happy to delete the analyzer if someone can let me know how to configure the test.

**Comment by mavasani** [comment]
> But I am happy to delete the analyzer if someone can let me know how to configure the test.

@CollinAlpert You can move the analyzer over to the test project and use it only for the tests. Add a comment + open a tracking issue to remove this analyzer in the future.

Please do make sure that the code fixer and this test-only analyzer use the Diagnostic ID that will be used in the Obsolete attribute on the APIs - I am not sure if a CAxxxx diagnostic ID is appropriate to indicate obsoletion, maybe they should use a special prefix, I see we have https://learn.microsoft.com/en-us/dotnet/fundamentals/syslib-diagnostics/obsoletions-overview#reference. I think you should first confirm the diagnostic ID in the issue and ensure it doesn't get used by another feature.

**Comment by CollinAlpert** [comment]
@mavasani What's missing here? I received no response in the issue, so I am assuming the diagnostic ID is okay. Is there anything else I need to do before this can be merged?

**Comment by buyaa-n** [comment]
@mavasani @stephentoub if the fixer should be fired for Obsoletion diagnostic ID is this repo is the correct place for the fixer? As far as I know most SYSLIB**** analyzer/fixers implemented in runtime, but I think the obsoletion analyzer is in roslyn repo.

> What's missing here? I received no response in the issue, so I am assuming the diagnostic ID is okay. Is there anything else I need to do before this can be merged?

Sorry for delay @CollinAlpert, we can's assume the diagnostic ID used here is okay, anybody might use that ID for obsoleting other API (the rule is just grab next available ID). You would need assign the ID for this obsoletion and mark the related APIs with that check this https://github.com/dotnet/runtime/pull/84812/files for example. But we might not want to mark the APIs until the fixer is implemented, not sure, what you think @stephentoub?

**Comment by buyaa-n** [comment]
Actually with https://github.com/dotnet/roslyn-sdk/issues/1117 being fixed seems now we can add the fixer here.

**Comment by CollinAlpert** [comment]
So I would need to create a PR for dotnet/runtime first, adding the `[Obsolete]` attributes?

**Comment by mpidash** [comment]
> Actually with [dotnet/roslyn-sdk#1117](https://github.com/dotnet/roslyn-sdk/issues/1117) being fixed seems now we can add the fixer here. Without having an anlyzer for testing only.

@CollinAlpert I pushed my WIP implementation of a fixer that uses no additional analyzer (with the fix from Roslyn SDK), see here: https://github.com/mpidash/roslyn-analyzers/commit/39276c6a80c002dacafabf0071c9f1b090906001#diff-b057b03b5cca1496dc3d962dd2593bfefe06119259299f588966b2c344a78ddeR20-R73
I also only used the IOperation API instead of checking the syntax in language specific fixers. Not sure which is preferred though.

Feel free to copy anything you need.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Usage/UseVolatileReadWrite.Fixer.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var root = await context.Document.GetRequiredSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var node = root.FindNode(context.Span, getInnermostNodeForTie: true);
+            if (TryGetThreadVolatileReadWriteMemberAccess(node, ThreadVolatileReadMethodName, out var readAccess))
+            {
+                var codeAction = CodeAction.Create(
+                    MicrosoftNetCoreAnalyzersResources.UseVolatileReadTitle,
+                    _ => Task.FromResult(context.Document.WithSyntaxRoot(root.ReplaceNode(readAccess, CreateVolatileMemberAccess(context.Document, VolatileReadMethodName)))),
+                    MicrosoftNetCoreAnalyzersResources.UseVolatileReadTitle
```
Comment (mpidash) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Usage/UseVolatileReadWrite.Fixer.cs:31]: Nit (see https://github.com/dotnet/roslyn-analyzers/issues/5431):
```suggestion
                    nameof(MicrosoftNetCoreAnalyzersResources.UseVolatileReadTitle)
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Usage/UseVolatileReadWrite.Fixer.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+                );
+                context.RegisterCodeFix(codeAction, context.Diagnostics);
+            }
+            else if (TryGetThreadVolatileReadWriteMemberAccess(node, ThreadVolatileWriteMethodName, out var writeAccess))
+            {
+                var codeAction = CodeAction.Create(
+                    MicrosoftNetCoreAnalyzersResources.UseVolatileWriteTitle,
+                    _ => Task.FromResult(context.Document.WithSyntaxRoot(root.ReplaceNode(writeAccess, CreateVolatileMemberAccess(context.Document, VolatileWriteMethodName)))),
+                    MicrosoftNetCoreAnalyzersResources.UseVolatileWriteTitle
```
Comment (mpidash) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Usage/UseVolatileReadWrite.Fixer.cs:40]: Nit (see https://github.com/dotnet/roslyn-analyzers/issues/5431):

```suggestion
                    nameof(MicrosoftNetCoreAnalyzersResources.UseVolatileWriteTitle)
```

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Usage/CSharpUseVolatileReadWriteFixer.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.NetCore.Analyzers.Usage;
+
+namespace Microsoft.NetCore.CSharp.Analyzers.Usage
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp), Shared]
+    internal sealed class CSharpUseVolatileReadWriteFixer : UseVolatileReadWriteFixer
+    {
+        protected override bool TryGetThreadVolatileReadWriteMemberAccess(SyntaxNode invocation, string methodName, [NotNullWhen(true)] out SyntaxNode? memberAccess)
```
Comment (mpidash) [inline src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Usage/CSharpUseVolatileReadWriteFixer.cs:15]:
```suggestion
        protected sealed override bool TryGetThreadVolatileReadWriteMemberAccess(SyntaxNode invocation, string methodName, [NotNullWhen(true)] out SyntaxNode? memberAccess)
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Usage/UseVolatileReadWrite.Fixer.cs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var root = await context.Document.GetRequiredSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var node = root.FindNode(context.Span, getInnermostNodeForTie: true);
+            if (TryGetThreadVolatileReadWriteMemberAccess(node, ThreadVolatileReadMethodName, out var readAccess))
+            {
+                var codeAction = CodeAction.Create(
+                    MicrosoftNetCoreAnalyzersResources.UseVolatileReadTitle,
+                    _ => Task.FromResult(context.Document.WithSyntaxRoot(root.ReplaceNode(readAccess, CreateVolatileMemberAccess(context.Document, VolatileReadMethodName)))),
```
Comment (mpidash) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Usage/UseVolatileReadWrite.Fixer.cs:30]: I think this breaks with named arguments, as `Volatile.Read` and `Thread.VolatileRead` use different parameter names (`location` and `address`).

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Usage/UseVolatileReadWrite.Fixer.cs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        private static SyntaxNode CreateVolatileMemberAccess(Document document, string methodName)
+        {
+            var generator = SyntaxGenerator.GetGenerator(document);
+            return generator.MemberAccessExpression(
+                generator.IdentifierName(nameof(Volatile)),
+                generator.IdentifierName(methodName)
+            );
+        }
+
+        public override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;
```
Comment (mpidash) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Usage/UseVolatileReadWrite.Fixer.cs:55]:
```suggestion
        public sealed override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Usage/UseVolatileReadWrite.Fixer.cs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+        protected abstract bool TryGetThreadVolatileReadWriteMemberAccess(SyntaxNode invocation, string methodName, [NotNullWhen(true)] out SyntaxNode? memberAccess);
+
+        public override ImmutableArray<string> FixableDiagnosticIds { get; } = ImmutableArray.Create("SYSLIB0054");
```
Comment (mpidash) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Usage/UseVolatileReadWrite.Fixer.cs:59]:
```suggestion
        public sealed override ImmutableArray<string> FixableDiagnosticIds { get; } = ImmutableArray.Create("SYSLIB0054");
```

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Usage/UseVolatileReadWriteTests.cs [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+using VerifyCS = Test.Utilities.CSharpCodeFixVerifier<
+    Microsoft.NetCore.Analyzers.Usage.UnitTests.UseVolatileReadWriteAnalyzer,
+    Microsoft.NetCore.CSharp.Analyzers.Usage.CSharpUseVolatileReadWriteFixer>;
+using VerifyVB = Test.Utilities.VisualBasicCodeFixVerifier<
+    Microsoft.NetCore.Analyzers.Usage.UnitTests.UseVolatileReadWriteAnalyzer,
+    Microsoft.NetCore.VisualBasic.Analyzers.Usage.BasicUseVolatileReadWriteFixer>;
+
+namespace Microsoft.NetCore.Analyzers.Usage.UnitTests
+{
+    public sealed class UseVolatileReadWriteTests
```
Comment (mpidash) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Usage/UseVolatileReadWriteTests.cs:16]: Tests checking if trivia and named arguments are preserved are missing.

**Comment by buyaa-n** [comment]
> So I would need to create a PR for dotnet/runtime first, adding the [Obsolete] attributes?

In general yes, that includes adding the attributes to the APIs with the desired ID and message, find all usages and fix them, though having a fixer could help with fixing if there are many usages. Though, I would say it can be optional, add the attributes if you would like to, or continue implementing the fixer with `SYSLIB0054` ID. In case it is used for another obsoletion we could change it later, no big deal.

**Comment by buyaa-n** [comment]
> I pushed my WIP implementation of a fixer that uses no additional analyzer (with the fix from Roslyn SDK)

Thank you very much @mpidash, implementing a fixer that uses no additional analyzer is the what we need here

> I also only used the IOperation API instead of checking the syntax in language specific fixers. Not sure which is preferred though.

Definitely `IOperation` APIs should be used whenever possible, only parts that is not supported with `IOperation` should use language specific implementations

**Comment by CollinAlpert** [comment]
> Definitely IOperation APIs should be used whenever possible

@buyaa-n But doesn't `IOperation` require a semantic model and thus have a large performance overhead?

**Comment by CollinAlpert** [comment]
@mpidash Thanks a lot, your WIP was extremely helpful! :)

**Comment by buyaa-n** [comment]
> @buyaa-n But doesn't `IOperation` require a semantic model and thus have a large performance overhead?

I don't really know the performance impact, but I would not expect perf overhead, @mavasani @sharwell could reply to that. I know that `IOperation` APIs are newer and added to avoid language specific implementations. Writing one code and using for all languages is better at least for maintenance, code reuse perspective.

**Comment by mavasani** [comment]
> > @buyaa-n But doesn't `IOperation` require a semantic model and thus have a large performance overhead?
> 
> I don't really know the performance impact, but I would not expect perf overhead, @mavasani @sharwell could reply to that. I know that `IOperation` APIs are newer and added to avoid language specific implementations. Writing one code and using for all languages is better at least for maintenance, code reuse perspective.

There should be no negative performance impact on using IOperation APIs in analyzers and fixers. We already have a whole ecosystem of IOperation based analyzers and fixers, and they all share the same IOperation trees.

**Comment by CollinAlpert** [comment]
I switched to using `IOperation` APIs, however I wasn't able to make named arguments work without language-specific fixers.

I also had to disable the test `VB_UseVolatileWrite_WithReversedArguments`, since Visual Basic's arguments are not passed to the fixer in the order they are supplied to the method. I figure it is fine for the user to have to switch the order of the arguments themselves. Besides, the fixed code will still compile.

**Comment by buyaa-n** [comment]
> I switched to using `IOperation` APIs, however I wasn't able to make named arguments work without language-specific fixers.

Thanks, I believe you can use this extension method for that:
https://github.com/dotnet/roslyn-analyzers/blob/7b1f39d7b8bd81d003898c49d0320682b6f06c54/src/Utilities/Compiler/Extensions/IOperationExtensions.cs#L884-L888

**Comment by CollinAlpert** [comment]
@buyaa-n You mean when someone calls `Thread.VolatileRead(value: value, address: ref address)` the call should be rewritten to `Volatile.Read(location: address, value: value)` or even `Volatile.Read(address, value)`?

**Comment by buyaa-n** [comment]
Yes, replacing with `Volatile.Read(location: ref address, value: value)` is completely acceptable, if that is not doable, I am OK with  `Volatile.Read(ref address, value)` too.

**Comment by CollinAlpert** [comment]
@buyaa-n Done. Unfortunately I still need the language-specific fixers, since `Thread.VolatileWrite`'s and `Volatile.Write`'s argument names are different (`address` vs. `location`). Same with `Thread.VolatileWrite` and `Volatile.Read`.

**Comment by buyaa-n** [comment]
> @buyaa-n Done. Unfortunately I still need the language-specific fixers, since `Thread.VolatileWrite`'s and `Volatile.Write`'s argument names are different (`address` vs. `location`). Same with `Thread.VolatileWrite` and `Volatile.Read`.

OK, then `always populate the arguments in order` was not needed, I reverted it and enabled the VB test for reverse order to validate the fix in order

**Review by buyaa-n — APPROVED** [review]
LGTM, thank you @CollinAlpert

**Pull request merged**